File: iosApp/iosApp/BackendUserProfileRepository.swift
import Foundation
import FirebaseFirestore

class BackendUserProfileRepository {
    private let db = Firestore.firestore()

    func observeOccupant(occupantDocId: String, onChange: @escaping ([String: Any]?) -> Void) -> ListenerRegistration {
        return db.collection("Occupants").document(occupantDocId)
            .addSnapshotListener { snapshot, _ in
                onChange(snapshot?.data())
            }
    }

    func observeIsEnabled(uid: String, onChange: @escaping (Bool) -> Void) -> ListenerRegistration {
        return db.collection("Users").document(uid)
            .addSnapshotListener { snapshot, _ in
                let enabled = snapshot?.data()?["Enabled"] as? Bool ?? true
                onChange(enabled)
            }
    }

    func isUserEnabled(uid: String) async throws -> Bool {
        let userDoc = try await db.collection("Users").document(uid).getDocument()
        guard let data = userDoc.data() else { return false }
        return data["Enabled"] as? Bool ?? false
    }

    func updateFcmToken(uid: String, token: String) async throws {
        try await db.collection("Users").document(uid).updateData(["fcm_token": token])
    }

    func getProfile(uid: String) async throws -> OccupantProfile {
        let userDoc = try await db.collection("Users").document(uid).getDocument()
        guard userDoc.exists, let userData = userDoc.data() else {
            throw NSError(
                domain: "UserProfile",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "User profile not found. Please contact the admin."]
            )
        }
        let role = userData["Role"] as? String ?? ""
        let enabled = userData["Enabled"] as? Bool ?? false

        let occupantQuery = try await db.collection("Occupants").whereField("authUid", isEqualTo: uid).getDocuments()
        guard let occupantDoc = occupantQuery.documents.first else {
            throw NSError(
                domain: "UserProfile",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "Occupant profile not found. Please contact the admin."]
            )
        }
        let occupantData = occupantDoc.data()
        let occupantFromTimestamp = occupantData["OccupantFrom"] as? Timestamp

        return OccupantProfile(
            uid: uid,
            name: occupantData["Name"] as? String ?? "",
            email: occupantData["Email"] as? String ?? "",
            role: role,
            enabled: enabled,
            empId: occupantData["EMPID"] as? String ?? "",
            flatNumber: occupantData["FlatNumber"] as? String ?? "",
            occupantFrom: occupantFromTimestamp?.dateValue(),
            isCoordinator: occupantData["isCoordinator"] as? Bool ?? false,
            occupantDocId: occupantDoc.documentID,
            flatId: occupantData["flatId"] as? String ?? ""
        )
    }
}

File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendUserProfileRepository.kt
package com.humblesolutions.indsphinx.repository

import com.google.firebase.firestore.FirebaseFirestore
import com.humblesolutions.indsphinx.model.OccupantProfile
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

class BackendUserProfileRepository : UserProfileRepository {
    private val db = FirebaseFirestore.getInstance()

    suspend fun isUserEnabled(uid: String): Boolean {
        val userDoc = db.collection("Users").document(uid).get().await()
        return userDoc.getBoolean("Enabled") ?: false
    }

    fun observeOccupant(occupantDocId: String): Flow<Map<String, Any>?> = callbackFlow {
        val registration = db.collection("Occupants").document(occupantDocId)
            .addSnapshotListener { snapshot, _ ->
                trySend(snapshot?.data)
            }
        awaitClose { registration.remove() }
    }

    fun observeIsEnabled(uid: String): Flow<Boolean> = callbackFlow {
        val registration = db.collection("Users").document(uid)
            .addSnapshotListener { snapshot, _ ->
                val enabled = snapshot?.getBoolean("Enabled") ?: true
                trySend(enabled)
            }
        awaitClose { registration.remove() }
    }

    suspend fun updateFcmToken(uid: String, token: String) {
        db.collection("Users").document(uid).update("fcm_token", token).await()
    }

    override suspend fun getProfile(uid: String): OccupantProfile {
        val userDoc = db.collection("Users").document(uid).get().await()
        if (!userDoc.exists()) throw Exception("User profile not found. Please contact the admin.")
        val role = userDoc.getString("Role") ?: ""
        val enabled = userDoc.getBoolean("Enabled") ?: false

        val occupantQuery = db.collection("Occupants").whereEqualTo("authUid", uid).get().await()
        val occupantDoc = occupantQuery.documents.firstOrNull()
            ?: throw Exception("Occupant profile not found. Please contact the admin.")

        return OccupantProfile(
            uid = uid,
            name = occupantDoc.getString("Name") ?: "",
            email = occupantDoc.getString("Email") ?: "",
            role = role,
            enabled = enabled,
            empId = occupantDoc.getString("EMPID") ?: "",
            flatNumber = occupantDoc.getString("FlatNumber") ?: "",
            occupantFrom = occupantDoc.getTimestamp("OccupantFrom")?.toDate()?.time ?: 0L,
            isCoordinator = occupantDoc.getBoolean("isCoordinator") ?: false,
            occupantDocId = occupantDoc.id,
            flatId = occupantDoc.getString("flatId") ?: ""
        )
    }
}
